copy_shard.c

```diff
@@ -0,0 +1,139 @@
+#include "utils/relcache.h"
+#include "utils/tuplestore.h"
+
+
+/* local function forward declarations */
+static bool CopyDataFromTupleStoreToRelation(Tuplestorestate *tupleStore,
+											 Relation relation);
+
+
+/* declarations for dynamic loading */
+PG_FUNCTION_INFO_V1(worker_copy_shard_placement);
+
+
+/*
+ * worker_copy_shard_placement implements a internal UDF to copy a table's data from
+ * a healthy placement into a receiving table on an unhealthy placement. This
+ * function returns a boolean reflecting success or failure.
+ */
+Datum
+worker_copy_shard_placement(PG_FUNCTION_ARGS)
+{
+	text *shardRelationNameText = PG_GETARG_TEXT_P(0);
+	text *nodeNameText = PG_GETARG_TEXT_P(1);
+	int32 nodePort = PG_GETARG_INT32(2);
+	Oid shardRelationId = ResolveRelationId(shardRelationNameText);
+	char *shardRelationName = text_to_cstring(shardRelationNameText);
+	char *nodeName = text_to_cstring(nodeNameText);
+
+
+	Relation shardTable = heap_open(shardRelationId, RowExclusiveLock);
+	ShardPlacement *placement = (ShardPlacement *) palloc0(sizeof(ShardPlacement));
+	Task *task = (Task *) palloc0(sizeof(Task));
+	StringInfo selectAllQuery = makeStringInfo();
+
+	TupleDesc tupleDescriptor = RelationGetDescr(shardTable);
+	Tuplestorestate *tupleStore = tuplestore_begin_heap(false, false, work_mem);
+	bool fetchSuccessful = false;
+	bool loadSuccessful = false;
+
+	appendStringInfo(selectAllQuery, SELECT_ALL_QUERY,
+	                 quote_identifier(shardRelationName));
+
+	placement->nodeName = nodeName;
+	placement->nodePort = nodePort;
+
+	task->queryString = selectAllQuery;
+	task->taskPlacementList = list_make1(placement);
+
+	fetchSuccessful = ExecuteTaskAndStoreResults(task, tupleDescriptor, tupleStore);
+	if (!fetchSuccessful)
+	{
+		ereport(ERROR, (errmsg("could not receive query results")));
+	}
+
+	loadSuccessful = CopyDataFromTupleStoreToRelation(tupleStore, shardTable);
+	if (!loadSuccessful)
+	{
+		ereport(ERROR, (errmsg("could not load query results")));
+	}
+
+	tuplestore_end(tupleStore);
+
+	heap_close(shardTable, RowExclusiveLock);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * CopyDataFromTupleStoreToRelation loads a specified relation with all tuples
+ * stored in the provided tuplestore. This function assumes the relation's
+ * layout (TupleDesc) exactly matches that of the provided tuplestore. This
+ * function returns a boolean indicating success or failure.
+ */
+static bool
+CopyDataFromTupleStoreToRelation(Tuplestorestate *tupleStore, Relation relation)
+{
+	TupleDesc tupleDescriptor = RelationGetDescr(relation);
+	TupleTableSlot *tupleTableSlot = MakeSingleTupleTableSlot(tupleDescriptor);
+	bool copySuccessful = false;
+
+	for (;;)
+	{
+		HeapTuple tupleToInsert = NULL;
+		Oid insertedOid = InvalidOid;
+		bool nextTuple = tuplestore_gettupleslot(tupleStore, true, false, tupleTableSlot);
+		if (!nextTuple)
+		{
+			copySuccessful = true;
+			break;
+		}
+
+		tupleToInsert = ExecMaterializeSlot(tupleTableSlot);
```

Comment:
`ExecMaterializeSlot` _does_ allocate memory and it first appears that this means we need to call `heap_freetuple` to avoid memory problems here; however, it also sets `tupleTableSlot->tts_shouldFree` to `true`, which causes `ExecClearTuple` to call `pfree` on `tupleTableSlot->tts_tuple`, i.e. the tuple.

So in short, there is no memory problem here. We load a (non-copy) tuple into the slot, create a copy (the heap tuple) to do the insert, then free it (because the flag set on the slot causes the clear call to free the associated tuple).